### ISSUE TITLE
[REVIEW]: Basic Threading Examples in JuliaLang v1.3

### ISSUE THREAD

**Issue by whedon** [JuliaCon/proceedings-review#54]
**Submitting author:** @vtjnash (<a href="http://orcid.org/0000-0002-8991-0523">Jameson Nash</a>)
**Repository:** <a href="https://github.com/vtjnash/JuliaCon2019Threading" target ="_blank">https://github.com/vtjnash/JuliaCon2019Threading</a>
**Editor:** @vchuravy
**Reviewer:** @ArchRobison, @wsmoses 
**Archive:** <a href="https://doi.org/10.5281/zenodo.4480494" target="_blank">10.5281/zenodo.4480494</a>

## Status

[![status](https://proceedings.juliacon.org/papers/9ac8bda44df0efa04afdfba84cb90c9e/status.svg)](https://proceedings.juliacon.org/papers/9ac8bda44df0efa04afdfba84cb90c9e)

Status badge code:

```
HTML: <a href="https://submissions.juliacon.org/papers/9ac8bda44df0efa04afdfba84cb90c9e"><img src="https://submissions.juliacon.org/papers/9ac8bda44df0efa04afdfba84cb90c9e/status.svg"></a>
Markdown: [![status](https://submissions.juliacon.org/papers/9ac8bda44df0efa04afdfba84cb90c9e/status.svg)](https://submissions.juliacon.org/papers/9ac8bda44df0efa04afdfba84cb90c9e)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://www.overleaf.com/read/bwwjvvmvzrdn" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@ArchRobison & @wsmoses , please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/JuliaCon/proceedings-reviews/invitations

The reviewer guidelines are available here: https://proceedings.juliacon.org/guide/reviewers. Any questions/concerns please let @vchuravy know.

✨ **Please try and complete your review in the next two weeks** ✨

### Conflict of interest

- [ ] As the reviewer I confirm that I have read the [JuliaCon conflict of interest policy](https://proceedings.juliacon.org/guide/authors#coi) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JuliaCon code of conduct](https://proceedings.juliacon.org/coc).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://www.overleaf.com/read/bwwjvvmvzrdn">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Authorship:** Has the submitting author (@vtjnash) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Paper format

- [ ] **Authors:** Does the `paper.tex` file include a list of authors with their affiliations?
- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [ ] **Page limit:** Is the page limit for full papers respected by the submitted document?

### Content

- [ ] **Context:** is the scientific context motivating the work correctly presented?
- [ ] **Methodology:** is the approach taken in the work justified, presented with enough details and reference to reproduce it?
- [ ] **Results:** are the results presented and compared to approaches with similar goals?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JuliaCon conflict of interest policy](https://proceedings.juliacon.org/guide/authors#coi) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JuliaCon code of conduct](https://proceedings.juliacon.org/coc).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://www.overleaf.com/read/bwwjvvmvzrdn">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Authorship:** Has the submitting author (@vtjnash) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Paper format

- [x] **Authors:** Does the `paper.tex` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [ ] **Page limit:** Is the page limit for full papers respected by the submitted document?

### Content

- [ ] **Context:** is the scientific context motivating the work correctly presented?
- [ ] **Methodology:** is the approach taken in the work justified, presented with enough details and reference to reproduce it?
- [ ] **Results:** are the results presented and compared to approaches with similar goals?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @ArchRobison, @wsmoses  it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/JuliaCon/proceedings-review) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/JuliaCon/proceedings-review:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #54 with the following error: 

 /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/find.rb:43:in `block in find': No such file or directory - tmp/54 (Errno::ENOENT)
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/find.rb:43:in `collect!'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/find.rb:43:in `find'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/lib/whedon/processor.rb:61:in `find_paper_paths'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/bin/whedon:50:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by vchuravy** [comment]
X-ref https://github.com/JuliaCon/proceedings-review/issues/49#issuecomment-534358218

**Comment by vchuravy** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by vchuravy** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to accept the paper and deposit with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00054/jcon.00054/10.21105.jcon.00054.pdf)

**Comment by vtjnash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00054/jcon.00054/10.21105.jcon.00054.pdf)

**Comment by vtjnash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00054/jcon.00054/10.21105.jcon.00054.pdf)

**Comment by vtjnash** [comment]
Is there anything else I need to do? I think I've adopted and incorporated all of Arch's feedback and made it better.

**Comment by vtjnash** [comment]
Am I supposed to bump this? I'm not really sure what its status is or who it's waiting on.

**Comment by wsmoses** [comment]
@vtjnash Do you need to @sync in the merge sort example before merging or am I missing something.

**Comment by vtjnash** [comment]
No, that's implied in the `fetch`

**Comment by wsmoses** [comment]
Ah that makes sense, in any case here's a more full review/set of comments:

This paper presents a mechanism of embedding dynamic task-based parallelism into the Julia language, allowing Julia programs to benefit from parallelism as distinct from existing Julia mechanisms from concurrency. The core of this contribution is a new scheduler with frontend modifications that allow several front-end improvements to make it easier to use this model. The authors also describe several changes to language constructs needed to integrate their framework with arbitrary code, including a demonstration of using external libraries with their framework.

The authors present several demonstrations of programs (merge sort, number sieve, and a prefix scan), that could benefit from the task-parallelism that they now permit. The additional syntax, while demonstrated in code blocks, isn't explained well for new users hoping to take advantage of it (e.g. no description of @sync/fetch, ambiguity about the guarantees provided by @spawn).

The authors allude to several limitations in existing Julia threading primitives. There isn't a good description of why the existing threading primitives couldn't be used on the demonstration codes (though I can see why this would be difficult). It would also be useful to do a performance comparison between the new runtime and old systems on some mutually compatible code. This comparison, along with perhaps showing the serial runtime [e.g. without parallel constructs rather than 1-core with parallel constructs] would be helpful to demonstrate that the new runtime doesn't provide significant overhead.

It would also be desirable to provide more quantifiable justification of their performance on FFTW than simply claiming it to be competitive (which is somewhat unclear). The authors should also include additional reference to related work in the field (e.g. existing parallel runtimes, including those outside of Julia) and describe the similarities and differences between prior art and the runtime described in this paper.

As a minor point (and I'm not the most familiar with the types of papers in this venue so take this with a grain of salt): some of the wording e.g. "but on to even more fun stuff..." could be tightened up for a professional audience.

Overall this paper represents a significant contribution to the Julia community in improving the experience of writing high performance parallel code. There are some additional changes to the writeup that would aid in fully demonstrating this contribution.

**Comment by vchuravy** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1007/978-3-319-20119-1_10 is OK
- 10.1145/2807591.2807602 is OK
- 10.1145/1248377.1248396 is OK
- 10.1145/359576.359585 is OK
- 10.1145/361020.361216 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by vchuravy** [comment]
Thank you @wsmoses for your review.

@vtjnash thank you for your patience can you address the reviewer comments and is there an archival DOI that you would like to associate with the publication?

**Comment by vtjnash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00054/jcon.00054/10.21105.jcon.00054.pdf)

**Comment by vtjnash** [comment]
I don't have an archival DOI, but don't think it'll be useful for this. I think Kiran and I have improved the paper based on the reviewer's comments.

**Comment by vtjnash** [comment]
What's next for us?

**Comment by vchuravy** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# All commands can be run on a non-default branch, to do this pass a custom 
# branch name by following the command with `from branch custom-branch-name`.
# For example:

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a dry run of accepting the paper and depositing with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by vchuravy** [comment]
@whedon accept

**Comment by whedon** [comment]
No archive DOI set. Exiting...

**Comment by vchuravy** [comment]
>  don't have an archival DOI, but don't think it'll be useful for this.

Can you create one with the example code being used?

**Comment by vchuravy** [comment]
@whedon remind @vtjnash in 1 week

**Comment by whedon** [comment]
Reminder set for @vtjnash in 1 week

**Comment by vtjnash** [comment]
@whedon set 10.5281/zenodo.4480494 as archive
@whedon generate pdf

**Comment by whedon** [comment]
I'm sorry @vtjnash, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by vchuravy** [comment]
@whedon set 10.5281/zenodo.4480494 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4480494" target="_blank">10.5281/zenodo.4480494</a> is the archive.

**Comment by vchuravy** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00054/jcon.00054/10.21105.jcon.00054.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00054/jcon.00054/10.21105.jcon.00054.pdf) :page_facing_up: :point_left:

**Comment by vchuravy** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-20119-1_10 is OK
- 10.1145/2807591.2807602 is OK
- 10.1145/1248377.1248396 is OK
- 10.1145/359576.359585 is OK
- 10.1145/361020.361216 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by vchuravy** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-20119-1_10 is OK
- 10.1145/2807591.2807602 is OK
- 10.1145/1248377.1248396 is OK
- 10.1145/359576.359585 is OK
- 10.1145/361020.361216 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @JuliaCon/jcon-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/JuliaCon/proceedings-papers/pull/37

If the paper PDF and Crossref deposit XML look good in https://github.com/JuliaCon/proceedings-papers/pull/37, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by vchuravy** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JCON!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/JuliaCon/proceedings-papers/pull/38
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/jcon.00054](https://doi.org/10.21105/jcon.00054)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://proceedings.juliacon.org/papers/10.21105/jcon.00054/status.svg)](https://doi.org/10.21105/jcon.00054)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/jcon.00054">
  <img src="https://proceedings.juliacon.org/papers/10.21105/jcon.00054/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://proceedings.juliacon.org/papers/10.21105/jcon.00054/status.svg
   :target: https://doi.org/10.21105/jcon.00054
```
This is how it will look in your documentation:

[![DOI](https://proceedings.juliacon.org/papers/10.21105/jcon.00054/status.svg)](https://doi.org/10.21105/jcon.00054)

**We need your help!**

JuliaCon Proceedings is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://forms.gle/yANMe4c9YmuW9MWcA
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-jcon
